### PR TITLE
Refactor iterator

### PR DESCRIPTION
**Pull request by jvillafanez**
## Description
Replaces the old file iterator with a new one. List of changes includes:
* Move related iterator classes to its own namespace (under Utilities)
* Provide a way to include more conditions over the iterator if more are needed in the future (no additional refactor should be needed if this happens, just minor code changes)
* Provide a iterator factory, currently focused on what we need at the moment. More generic methods could be added in the future if required.
* Unittests over all the touched classes, with almost full coverage ("key" method on the RecursiveNodeIterator might need some eyes to check what to do as it doesn't work as expected. It shouldn't be needed though)

Supersedes https://github.com/owncloud/data_exporter/pull/31

## Related Issue
https://github.com/owncloud/data_exporter/issues/29

## How Has This Been Tested?
Unittest added. Manual tests not done yet.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

@IljaN Check if there is any additional explanation that you might need to understand the code. There is still the problem to know what to do about the "key" method of the iterator

### CODE REVIEW

**Comment by jvillafanez** [comment]
There might be problems with the export. I don't think the paths are being generated properly...

**Comment by jvillafanez** [comment]
Also checked with files_primary_s3 app as primary storage without any problem.

**Comment by IljaN** [comment]
> Also checked with files_primary_s3 app as primary storage without any problem.

Also with import? :dancer:

**Comment by jvillafanez** [comment]
> Also with import?

Versioning seems to be a problem, not sure if the files_versions app is the one to blame. The versions are in the oc_filecache table and also in the S3, but as a different file, not as version of the file.

File: phpstan.neon [hunk 1]
```diff
@@ -3,3 +3,5 @@ parameters:
   ignoreErrors:
     - '#If condition is always true#'
     # lib/Exporter/MetadataExtractor/SharesExtractor.php:132 (require changes in core)
+    - '#Strict comparison using === between string and null will always evaluate to false#'
+    # lib/Utilities/Iterators/Nodes/SkipNodeConditionIgnorePath.php:47 (require changes in core, mismatch between interface and implementation)
```
Comment (patrickjahns) [inline phpstan.neon]: Where is the ticket for this issue?

**Comment by jvillafanez** [comment]
Opened https://github.com/owncloud/core/issues/33047 and https://github.com/owncloud/core/issues/33048 for the 2 exclusions

Note that the line for the first exclusion has been shifted to line 159

**Comment by jvillafanez** [comment]
@patrickjahns anything to add? This should be ready to merge.

**Comment by patrickjahns** [comment]
Not able to do a full code review right now - so if you need a code review, please raise with engineering team

**Pull request merged**